# DS 3000 - Assignment 02

**Student Name**: Matthew Tong

**Date**: 1/18/25


### Submission Instructions
Submit this `ipynb` file to canvas.

The `ipynb` format stores outputs from the last time you ran the notebook.  (When you open a notebook it has the figures and outputs of the last time you ran it too).  To ensure that your submitted `ipynb` file represents your latest code, make sure to give a fresh run `Kernel > Restart & Run All` just before uploading the `ipynb` file to Canvas.

### Academic Integrity

**Writing your homework is an individual effort.**  You may discuss general python problems with other students but under no circumstances should you observe another student's code which was written for this assignment, from this year or past years.  Pop into office hours or DM us in MS Teams if you have a specific question about your work or if you would like another pair of eyes or talk through your code.

Don't forget to cite websites which helped you solve a problem in a unique way.  You can do this in markdown near the code or with a simple one-line comment. You do not need to cite the official python documentation.

**Documentation / style counts for credit**  Please refer to the Pep-8 style, to improve the readability and consistency of your Python code. For more information, read the following article [How to Write Beautiful Python Code With PEP 8](https://realpython.com/python-pep8/) or ask your TA's for tips.

**NOTE:<span style='color:red'> Write python expressions to answer ALL questions below and ensure that you use the `print()` function to display the output.</span>** Each question should be answered in a new code cell. For example, your solution for question 1.1 should be in a different code cell from your solution for question 1.2.

## Question 1.1: 
Load the attached data into a numpy array and display the dimensions of the data : birthwt.csv
Note: do not count the column headers, only the data.

In [245]:
import numpy as np

data = np.loadtxt("birthwt.csv", delimiter = ",", skiprows = 1)
print(data.shape)

(189, 10)


## Question 1.2: 
What is the age range of mothers in this dataset.
Note: the range should include: the minimum, average and maximum age. Ensure that the ages are represented as whole numbers.


In [248]:
ages = data[:,1]

minimum_age = int(np.min(ages))
average_age = int(np.mean(ages))
maximum_age = int(np.max(ages))

print("Minimum Age:", minimum_age)
print("Average Age:", average_age)
print("Maximum Age:", maximum_age)

Minimum Age: 14
Average Age: 23
Maximum Age: 45


## Question 1.3: 
How many mothers smoked during pregnancy?


In [251]:
mothers_smoked = data[:, 4]

num_smoked = int(np.sum(mothers_smoked))

print(num_smoked, "mothers smoked during pregnancy. ")

74 mothers smoked during pregnancy. 


## Question 1.4: 
How many mothers had a history of hypertension?


In [254]:
mothers_hypertension = data[:, 6]

num_hypertension = int(np.sum(mothers_hypertension))

print(num_hypertension, "mothers had a history of hypertension. ")

12 mothers had a history of hypertension. 


## Question 1.5: 
How many mothers have 1 or more premature births?


In [257]:
mothers_premature_birth = data[:, 5]

num_premature_birth = int(np.sum(mothers_premature_birth >= 1))

print(num_premature_birth, "mothers had 1 or more premature births. ")

30 mothers had 1 or more premature births. 


## Question 1.6: 
Create a function called convert_birthweight() that:
takes grams as input,
converts grams to the equivalent pounds and ounces
example format: X grams = Y lbs Z oz; note: X, Y and Z are placeholders for numeric values.
return the converted value.
Use the convert_birthweight() function to convert the birth weight to pounds and ounces in the questions 7-11 below.


In [260]:
OUNCES_PER_LB = 453.592
OUNCES_PER_GRAM = 28.3495

def convert_birthweight(grams): 
    lbs = int(grams // OUNCES_PER_LB)
    ounces = round((grams % OUNCES_PER_LB) / OUNCES_PER_GRAM, 2)
    return str(lbs) + " lbs " + str(ounces) + " ounces"

## Question 1.7: 
What is the average birth weight of babies whose mothers have a history of premature birth? 

In [263]:
birth_weight = data[:, 9]

avg_premature_weight = np.mean(birth_weight[mothers_premature_birth >= 1])
converted_weight = convert_birthweight(avg_premature_weight)

print("The average birth weight of babies whose mothers have a history of premature birth is", converted_weight)

The average birth weight of babies whose mothers have a history of premature birth is 5 lbs 10.99 ounces


## Question 1.8: 
What is the minimum, average and maximum birth weight of babies whose mother did not smoke during pregnancy and do not have a history of hypertension?

In [266]:
weight_mother_no_smoke_ht = birth_weight[(data[:, 4] == 0) & (data[:, 6] == 0)]

minimum_weight = np.min(weight_mother_no_smoke_ht)
average_weight = np.mean(weight_mother_no_smoke_ht)
maximum_weight = np.max(weight_mother_no_smoke_ht)

converted_min = convert_birthweight(minimum_weight)
converted_average = convert_birthweight(average_weight)
converted_max = convert_birthweight(maximum_weight)

print("Minimum Birth Weight", converted_min)
print("Average Birth Weight", converted_average)
print("Maximum Birth Weight", converted_max)

Minimum Birth Weight 2 lbs 4.01 ounces
Average Birth Weight 6 lbs 13.01 ounces
Maximum Birth Weight 11 lbs 0.02 ounces


## Question 1.9: 
What is the minimum, average and maximum birth weight of babies whose mothers smoked during pregnancy?

In [269]:
weight_mother_smoke = birth_weight[(data[:, 4] == 1)]

minimum_weight = np.min(weight_mother_smoke)
average_weight = np.mean(weight_mother_smoke)
maximum_weight = np.max(weight_mother_smoke)

converted_min = convert_birthweight(minimum_weight)
converted_average = convert_birthweight(average_weight)
converted_max = convert_birthweight(maximum_weight)

print("Minimum Birth Weight", converted_min)
print("Average Birth Weight", converted_average)
print("Maximum Birth Weight", converted_max)

Minimum Birth Weight 1 lbs 9.01 ounces
Average Birth Weight 6 lbs 1.78 ounces
Maximum Birth Weight 9 lbs 5.49 ounces


## Question 1.10: 
What is the minimum, average and maximum birth weight of babies whose mothers had a history of hypertension? 

In [272]:
weight_mother_hypertension = birth_weight[(data[:, 6] == 1)]

minimum_weight = np.min(weight_mother_hypertension)
average_weight = np.mean(weight_mother_hypertension)
maximum_weight = np.max(weight_mother_hypertension)

converted_min = convert_birthweight(minimum_weight)
converted_average = convert_birthweight(average_weight)
converted_max = convert_birthweight(maximum_weight)

print("Minimum Birth Weight", converted_min)
print("Average Birth Weight", converted_average)
print("Maximum Birth Weight", converted_max)

Minimum Birth Weight 2 lbs 8.04 ounces
Average Birth Weight 5 lbs 9.48 ounces
Maximum Birth Weight 8 lbs 5.69 ounces


## Question 1.11: 
What is the minimum, average and maximum birth weight of babies whose mothers smoked during pregnancy and had a history of hypertension? 

In [275]:
weight_mother_smoke_ht = birth_weight[(data[:, 4] == 1) & (data[:, 6] == 1)]

minimum_weight = np.min(weight_mother_smoke_ht)
average_weight = np.mean(weight_mother_smoke_ht)
maximum_weight = np.max(weight_mother_smoke_ht)

converted_min = convert_birthweight(minimum_weight)
converted_average = convert_birthweight(average_weight)
converted_max = convert_birthweight(maximum_weight)

print("Minimum Birth Weight", converted_min)
print("Average Birth Weight", converted_average)
print("Maximum Birth Weight", converted_max)

Minimum Birth Weight 2 lbs 8.04 ounces
Average Birth Weight 5 lbs 10.34 ounces
Maximum Birth Weight 8 lbs 4.49 ounces


## Question 1.12: 
Compare and contrast the results from 7-11 above and summarize your findings.
Do you see any patterns in the birth weight of babies whose mother did not have any history of hypertension and/or did not smoke during pregnancy versus those who did? Ensure that you reference specific birth weights to support your explanation.

One pattern that I found was that all the babies weighed roughly the same except for when the mother did not smoke or have hypertension. (5 lbs 10.34 ounces, 5 lbs 9.48 ounces, 6 lbs 1.78 ounces) vs 6 lbs 13.01 ounces. 

Another pattern that I saw was that babies whose mothers smoked/had hypertension had either a lower min birth weight or lower max birth weight. For example, no smoking and no hypertension babies were 2 lbs 4.01 ounces minimum. Mothers who smoked had a minimum of 1 lbs 9.01 ounces. For maximum, no smoking and no hypertension max was 9 lbs 5.49 ounces. Mothers who smoked/had hypertension or both had weights of 8 lbs 4.49 ounces, 8 lbs 5.69 ounces, and 8 lbs 5.69 ounces. 

Therefore, based on my observations, babies who come from mothers who did not smoke or have hypertension weighted more, which is based on my first point which highlights this. In addition, the second point shows that the minimums and maximums are lower when mothers smoke/have hypertension. 

## Helpful resources 
Don't forget to cite websites which helped you solve a problem in a unique way.  You can do this in markdown near the code or with a simple one-line comment inside the code cell, or you can list them below. 

You do not need to cite the official python documentation.


https://stackoverflow.com/questions/20457038/how-to-round-to-2-decimals-with-python
//used to round all numeric calculations (i.e. floats) to 2 decimal places